### PR TITLE
Scrape test

### PR DESCRIPTION
**Pull request by Hussein-Chukair**
This is just a scraper that works for some of the youtube ads. This isnt pushed to Cloud yet. It will be pushed once its been revised to work for all ads.

### CODE REVIEW

File: services/scrape-test/insert_youtube_url.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from google.cloud import bigquery
+from typing import Dict, Any
+import datetime
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def serialize_row(row: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (k-szewczyk) [inline services/scrape-test/insert_youtube_url.py:9]: This function name is misleading. It's not really doing serialization. it just reformats date. Also, maybe consider create data transfer object from it here using for example `pydantic` or `@dataclass`

File: services/scrape-test/scrape_media_url.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        url (str): The URL to scrape.
+        context (BrowserContext): The Playwright browser context to use for scraping.
+
+    Returns:
+        Optional[str]: The YouTube link if found, "timeout_error" on timeout, or None if no link is found.
+    """
+    logger.info(f"Starting to scrape for YouTube link at URL: {url}")
+    page = context.new_page()
+    try:
+        page.goto(url, timeout=60000)
```
Comment (k-szewczyk) [inline services/scrape-test/scrape_media_url.py:22]: I wouldn't use that big timeouts. if for some reason, that page is broken it just blocks your process. It's better to have shorter timeout and store unsuccessful url's somewhere, and try to go back to them again

File: services/scrape-test/scrape_media_url.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                    logger.info(f"Found YouTube iframe with src: {frame.url}")
+                    return frame.url
+                iframe_elements = frame.query_selector_all("iframe")
+                for iframe in iframe_elements:
+                    child_frame = iframe.content_frame()
+                    if child_frame:
+                        result = find_youtube_in_frames([child_frame])
+                        if result:
+                            return result
+            return None
```
Comment (k-szewczyk) [inline services/scrape-test/scrape_media_url.py:37]: I'd do more something like this:
```suggestion
        def find_youtube_in_frames(frames):
            for frame in frames:
                if "youtube.com" in frame.url:
                    logger.info(f"Found YouTube iframe with src: {frame.url}")
                    return frame.url
                iframe_elements = frame.query_selector_all("iframe")
                for iframe in iframe_elements:
                    child_frame = iframe.content_frame()
                    if not child_frame:
                        return None
                    return find_youtube_in_frames([child_frame])
            return None
```
(But test it first, I've didn't test if this works properly)
try to don't nest if's that many times, it makes code harder to read.

File: services/scrape-test/scrape_youtube_service.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                    row["youtube_video_url"] = youtube_link
+                    destination_table = (
+                        "annular-net-436607-t0.sample_ds.new_table_with_youtube_link"
+                    )
+                    insert_row_to_bq(row, destination_table)
+                elif youtube_link == "timeout_error":
+                    destination_table = "annular-net-436607-t0.sample_ds.ads_probably_with_a_youtube_link"
+                    insert_row_to_bq(row, destination_table)
+
+        browser.close()
```
Comment (k-szewczyk) [inline services/scrape-test/scrape_youtube_service.py:63]: First I'ld move this logic out of this function. Second, I'ld rather do this flow different

`POST: scrape_media_url` starts this scraping process on different thread and returns it's ID

`GET: scrape_result/{id}` returns scrape result for this ID if it's done or info that it's still working or error that this id doesn't exists

We can discuss that one later

File: services/scrape-test/test_samples.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+from decimal import Decimal  # If Decimal is used elsewhere
+
+
+def get_rows_from_bq():
+    # Initialize the BigQuery client
+    client = bigquery.Client()
+
+    # Query to retrieve rows from the successful scrapes table
+    query = """
+        SELECT * FROM `annular-net-436607-t0.dbt.testConsumption` LIMIT 500
```
Comment (k-szewczyk) [inline services/scrape-test/test_samples.py:14]: I'ld move this table name to some config file

File: services/scrape-test/test_samples.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            row[key] = value.isoformat()
+        elif isinstance(value, Decimal):
+            row[key] = float(value)
+        elif value is None:
+            row[key] = None  # BigQuery accepts NULL values
+        elif isinstance(value, bool):
+            row[key] = bool(value)
+        else:
+            # Convert other types to strings if necessary
+            row[key] = str(value) if not isinstance(value, (int, float, str)) else value
```
Comment (k-szewczyk) [inline services/scrape-test/test_samples.py:38]: definitely use `pydantic` here

File: services/scrape-test/test_samples.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+                # Search for iframes within the frame
+                iframe_elements = frame.query_selector_all("iframe")
+                for iframe in iframe_elements:
+                    child_frame = iframe.content_frame()
+                    if child_frame:
+                        result = find_youtube_in_frames([child_frame])
+                        if result:
+                            return result
+            return None
```
Comment (k-szewczyk) [inline services/scrape-test/test_samples.py:85]: Why it's repeated code from dev?

File: services/scrape-test/test_samples.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+                youtube_link = scrape_youtube_link(url, context)
+
+                if (
+                    youtube_link
+                    and youtube_link != "timeout_error"
+                    and "youtube.com" in youtube_link
+                ):
+                    if successful_scrapes >= 50:
+                        print(
+                            "Reached 50 successful scrapes. Skipping insertion into original table."
```
Comment (k-szewczyk) [inline services/scrape-test/test_samples.py:163]: as before move this whole function somewhere else. Also this scraping need some major refactor

**Review by k-szewczyk — CHANGES_REQUESTED** [review]

File: services/scrape-test/scrape_media_url.py (hunk 2 above)
Comment (Hussein-Chukair) [reply services/scrape-test/scrape_media_url.py]: I changed it to 20 seconds for now, after testing more we will take it down another level

File: services/scrape-test/scrape_media_url.py (hunk 3 above)
Comment (Hussein-Chukair) [reply services/scrape-test/scrape_media_url.py]: I tried this but it didnt work, it exited the loop after the first Iframe found even if there wasnt a media link in it.

File: services/scrape-test/test_samples.py (hunk 7 above)
Comment (Hussein-Chukair) [reply services/scrape-test/test_samples.py]: This file was only a test script and not meant to be used later on, it's now deleted.

File: services/scrape-test/scrape_youtube_service.py (hunk 4 above)
Comment (Hussein-Chukair) [reply services/scrape-test/scrape_youtube_service.py]: I changed this to work the way you mentioned. It was a very good approach!

File: services/scrape-test/test_samples.py (hunk 8 above)
Comment (Hussein-Chukair) [reply services/scrape-test/test_samples.py]: This file has been deleted now.

File: services/scrape-test/app/logging_config.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+logging.basicConfig(
+    level=config["logging"]["log_level"],
+    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
+    handlers=[logging.StreamHandler()],
+)
+
+logger = logging.getLogger("YouTubeScraper")
+"""
+Logger instance for the application.
+"""
```
Comment (k-szewczyk) [inline services/scrape-test/app/logging_config.py:15]: ⭐️

File: services/scrape-test/app/routes.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    Args:
+        advertiser_id (str): The ID of the advertiser to process.
+        background_tasks (BackgroundTasks): FastAPI background tasks.
+
+    Returns:
+        JobCreateResponse: Contains the job ID of the started job.
+    """
+    job_id = str(uuid.uuid4())
+    global_vars.job_statuses[job_id] = "Pending"
+    global_vars.active_jobs.add(job_id)
```
Comment (k-szewczyk) [inline services/scrape-test/app/routes.py:37]: not a fan of global variables. I'ld use some in memory storage like Redis or something similiar

**Review by k-szewczyk — APPROVED** [review]
Great Work!

File: services/scrape-test/app/routes.py (hunk 10 above)
Comment (k-szewczyk) [reply services/scrape-test/app/routes.py:37]: But your solution is okay, it's just matter of opinions :D

**Review by d0ndada — APPROVED** [review]
Looks good for me

**Pull request merged**